aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -862,6 +862,23 @@ Tensor max(const Tensor& self, const Tensor& other) {
   return at::maximum(self, other);
 }
 
+Tensor& fmax_out(Tensor& result, const Tensor& self, const Tensor& other) {
+    auto self_isnan = self.isnan();
+    auto other_isnan = other.isnan();
+    auto both_nan = self_isnan.logical_and(other_isnan);
+    return at::maximum_out(result,
+                           at::where(self_isnan == both_nan, self, other.to(self.dtype())),
+                           at::where(other_isnan == both_nan, other, self.to(other.dtype())));
+}
+
+Tensor fmax(const Tensor& self, const Tensor& other) {
+    auto self_isnan = self.isnan();
+    auto other_isnan = other.isnan();
+    auto both_nan = self_isnan.logical_and(other_isnan);
+    return at::maximum(at::where(self_isnan == both_nan, self, other.to(self.dtype())),
```

Comment:
This can be written more efficiently as: __other_isnan || (!self_isnan && self > other)__
```
auto mask = other.isnan().logical_or_(self.isnan().logical_not_().logical_and_(self > other));
return at::where(mask, self, other);
```
You'd have to do the type promotion beforehand which I think is preferable.

Reply:
You could do something similar to the out version, though torch.where does not have an out version so you'd have to copy the result.

Reply:
This is an interesting idea, but I think it'll be easier to write this using TensorIterator, which will handle broadcasting and type promotion and be much faster. See my comment below.

Reply:
I've improved the code to use TensorIterator instead.